plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -297,6 +298,16 @@ class ArtifactSaver(Stage):
             artifacts_to_save = []
             for declarative_content in batch:
                 for declarative_artifact in declarative_content.d_artifacts:
+                    if not declarative_artifact.digest_present:
+                        # Artifact may exist in the database,
+                        # but we did not know the digest before downloading
+                        digest_name = declarative_artifact.artifact.DIGEST_FIELDS[0]
+                        digest_value = getattr(declarative_artifact.artifact, digest_name)
+                        if digest_value:
+                            existing_artifact = Artifact.objects.filter(
```

Comment:
I think the QueryExistingArtifacts uses a batch implementation so using that one will also avoid a one-by-one DB implementation.